[mono] Move basic w32file* and w32process* functions to eventpipe

akoeplinger opened the pull request:
They are not used by the rest of the runtime and we only need a bare minimum of w32process and w32file in eventpipe (e.g. no file share/access logic).

Fix ep-test build and warnings, it was missing linking against libz and monoapi.

Fix host vs. target confusion in ep-rt-mono.c

akoeplinger committed:
[mono] Move basic w32file* and w32process* functions to eventpipe

They are not used by the rest of the runtime and we only need a bare minimum of w32process and w32file in eventpipe (e.g. no file share/access logic).

Fix ep-test build and warnings, it was missing linking against libz and monoapi.

Fix host vs. target confusion in ep-rt-mono.c

am11 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2224,7 +2483,7 @@ static const int64_t SECS_TO_NS = 1000000000;
 
 /* clock_gettime () is found by configure on Apple builds, but its only present from ios 10, macos 10.12, tvos 10 and watchos 3 */
-#if defined (HAVE_CLOCK_MONOTONIC) && (defined(TARGET_IOS) || defined(TARGET_OSX) || defined(TARGET_WATCHOS) || defined(TARGET_TVOS))
+#if defined (HAVE_CLOCK_MONOTONIC) && (defined(HOST_IOS) || defined(HOST_OSX) || defined(HOST_WATCHOS) || defined(HOST_TVOS))
```
Minimum supported macOS version is 10.15 in .NET 6, per https://github.com/dotnet/core/blob/main/release-notes/6.0/supported-os.md#macos.

akoeplinger replied:
yes HOST_DARWIN would work too, good point. I can change that in a separate cleanup since we have similar code elsewhere.

Regarding removing this altogether, we have https://github.com/dotnet/runtime/issues/58737 which tracks removing it from Android/iOS but last time we tried we hit various issues (https://github.com/dotnet/runtime/pull/62978) so it needs a bit more care.

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2069,6 +2082,252 @@ ep_rt_mono_rand_try_get_bytes (
+	char escape_chars[] = "\'\\";
+#endif
+
+	gboolean need_quote = FALSE;
```
I know we ported this code into EventPipe, but should we try to switch and use C99 types as much as possible in ported code. If we call function using Mono types, we should of course use matching types, but for other things, I think we should try to align with EventPipe code.

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2069,6 +2082,252 @@ ep_rt_mono_rand_try_get_bytes (
+	return res;
+}
+
+static void
+win32_io_interrupt_handler (void* ignored)
```
```suggestion
static
void
win32_io_interrupt_handler (void *ignored)
```

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.h:
```diff
@@ -354,6 +352,11 @@ extern mono_lazy_init_t _ep_rt_mono_os_cmd_line_init;
 extern ep_rt_spin_lock_handle_t _ep_rt_mono_config_lock;
+extern char * ep_rt_mono_get_managed_cmd_line (void);
+extern char * ep_rt_mono_get_os_cmd_line (void);
+extern ep_rt_file_handle_t ep_rt_mono_file_open_write(const ep_char8_t *path);
```
```suggestion
extern ep_rt_file_handle_t ep_rt_mono_file_open_write (const ep_char8_t *path);
```

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.h:
```diff
@@ -1407,13 +1383,12 @@ inline
-	ep_rt_utf16_string_free (path_utf16);
-
-	return file_handle;
+	if (res == INVALID_HANDLE_VALUE)
```
```suggestion
	return (res != INVALID_HANDLE_VALUE) ? res : INVALID_HANDLE_VALUE;
```

lateralusX commented on src/mono/mono/eventpipe/test/ep-buffer-tests.c:
```diff
@@ -135,10 +135,10 @@ load_buffer (
 		EventPipeEventPayload payload;
 		gchar *event_data = NULL;
-		size_t event_data_len = 0;
+		uint32_t event_data_len = 0;
```
Maybe keep size_t and cast when used as uint32_t?

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2069,6 +2082,252 @@ ep_rt_mono_rand_try_get_bytes (
+	// we only return the native host here since getting the full commandline is complicated and
+	// it's not super important to have the correct value since it'll only be used during startup
+	// until we have the managed commandline
+	return minipal_getexepath();
```
EventPipe assumes that the string returned by this method can correctly be freed using g_free (os_command_line_lazy_clean). Normally that is true since we use Mono API's to create the strings that will come from the same heap as g_free will free to, but when we now switch to minipal_getexepath, that is semi true, since that method uses regular C strdup to allocate string, so in theory, that could end up allocating from one heap and free on another, but most likely they will map to the same heap. If we are going to be accurate we should either return a string that is allocated using Mono API's that should be freed using g_free or update os_command_line_lazy_clean to use free, but that will break the pattern a little.

lateralusX reviewed (commented):
Overall I think it looks great, nice work!. Most comments are formatting and some around adapting ported code to EventPipe types where possible.

akoeplinger committed:
Formatting suggestions from code review

Co-authored-by: Johan Lorensson <lateralusx.github@gmail.com>

akoeplinger commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2069,6 +2082,252 @@ ep_rt_mono_rand_try_get_bytes (
+bool
+ep_rt_mono_file_write (ep_rt_file_handle_t handle, const uint8_t *buffer, uint32_t numbytes, uint32_t *byteswritten)
+{
+	bool res;
```
The original w32file-win32.c function didn't have it, but I can add it if you think it should be there.

lateralusX replied:
We have the check in none windows implementation, so I guess the same should hold on Windows, so we should have same validation logic there as well.

akoeplinger committed:
More code review feedback

Fix more formatting.

Move initialization of num_main_args after setting main_args so we can use it as a cheap guard value (not totally thread-safe but good enough for our purposes)

akoeplinger committed:
Addressing feedback

Move cmdline formatting back into object.c
Fix bug return INVALID_HANDLE_VALUE instead of null in ep_rt_file_open_write

lateralusX reviewed (approved):
LGTM!
